Make release build use conan deps where possible and hbb 4.0.1

tequdev opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

<!--
Please check [x] relevant options, delete irrelevant ones.

* If there is any impact to the public API methods (HTTP / WebSocket), please update https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md
  * Update API-CHANGELOG.md and add the change directly in this PR by pushing to your PR branch.
* libxrpl: See https://github.com/XRPLF/rippled/blob/develop/docs/build/depend.md
* Peer Protocol: See https://xrpl.org/peer-protocol.html
-->

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

tequdev committed:
Make release build use conan

tequdev committed:
delete remove old files script

tequdev committed:
remove unused cmake files

tequdev committed:
use dockerfile

tequdev committed:
fix for static build

dangell7 committed:
Merge branch 'dev' into conan-release-builder

tequdev committed:
update to gcc-11 llvm14, use ninja

tequdev committed:
fix boost version

tequdev committed:
simplify release-builder.sh

tequdev committed:
Merge branch 'dev' into conan-release-builder

tequdev committed:
remove checkpatterns workflow

tequdev committed:
Merge branch 'dev' into conan-release-builder

tequdev committed:
remove unnecessary flags

tequdev committed:
fix

tequdev commented:
As of the latest build of this PR (2025.7.9-HEAD+1952), I have confirmed that the following three commands work (with `xahaud —version`).

And in WSL(Ubuntu 20.04)

## Debian 10

```bash
docker run --rm --platform linux/amd64 debian:10 bash -c "apt-get update && apt-get install -y wget && wget -O xahaud https://build.xahau.tech/2025.7.9-HEAD%2B1952 && chmod +x xahaud && ./xahaud --version"
```

## Ubuntu 20.04

```bash
docker run --rm --platform linux/amd64 ubuntu:20.04 bash -c "apt-get update && apt-get install -y wget && wget -O xahaud https://build.xahau.tech/2025.7.9-HEAD%2B1952 && chmod +x xahaud && ./xahaud --version"
```

## Rocky Linux 8 (RHEL8 Compatible)

```bash
docker run --rm --platform linux/amd64 rockylinux:8 bash -c "yum install -y wget && wget -O xahaud https://build.xahau.tech/2025.7.9-HEAD%2B1952 && chmod +x xahaud && ./xahaud --version"
```

sublimator commented on Builds/CMake/deps/WasmEdge.cmake:
```diff
@@ -1,84 +1 @@
-  message(WARNING "xar library not found... (only important for mac builds)")
-endif()
-add_library (wasmedge::wasmedge ALIAS wasmedge)
+find_package(wasmedge REQUIRED)
```
Did we check using `ldd` that this is not using the conan shared library?

sublimator replied:
I feel like the libraries made available via the conan toolchain file would probably take precedence here?

tequdev replied:
It has been confirmed that wasmedge is used as a static lib in xahaud binaries created with release-builder.

WasmEdge.cmake is rewritten by build-core.sh. (This is an existing process.)
https://github.com/tequdev/xahaud/blob/0badfd3c9e16de5ef4a4f85143a5f18e0cad1780/build-core.sh#L30-L39

sublimator replied:
Oh, I thought I had seen something about copying over that file, but then when I looked before quickly I couldn't see it

Thanks

sublimator commented on release-builder.sh:
```diff
@@ -50,25 +49,152 @@ STATIC_CONTAINER=$(docker ps -a | grep $CONTAINER_NAME |wc -l)
+    cp /usr/lib64/llvm14/lib/liblld*.a /usr/local/lib/ && \
+    cd /tmp && rm -rf lld-* libunwind-*"
+
+# Build and install WasmEdge
```
https://github.com/tequdev/xahaud/pull/4

This would have satisfied my curiousity before

sublimator committed:
docs: add explanation for static WasmEdge build dependencies (#4)

- clarify why boost, llvm, and lld are installed separately
- explain that conan only provides shared library version
- note which deps come from packages vs source builds

sublimator commented on release-builder.sh:
```diff
@@ -50,25 +49,158 @@ STATIC_CONTAINER=$(docker ps -a | grep $CONTAINER_NAME |wc -l)
+
+  # Create inline Dockerfile with environment setup for build-full.sh
+  DOCKERFILE_CONTENT=$(cat <<'DOCKERFILE_EOF'
+FROM ghcr.io/phusion/holy-build-box:4.0.1-amd64
```
You said 4.0.2 is out ?

sublimator replied:
https://github.com/phusion/holy-build-box/releases/tag/rel-4.0.2

tequdev replied:
The latest version is 4.0.1, and it seems that the 4.0.2 has not been released yet.

https://github.com/phusion/holy-build-box/pkgs/container/holy-build-box

tequdev committed:
add ccache `compiler_check` option

sublimator commented on release-builder.sh:
```diff
@@ -50,25 +49,159 @@ STATIC_CONTAINER=$(docker ps -a | grep $CONTAINER_NAME |wc -l)
   rm -rf release-build;
   mkdir -p release-build;
 
+  docker volume create cache-volume
```
If this is running on an actions runner with persistent state, which I believe the self-hosted runners qualify as, if there's no volume it will succeed but if it's already existing it will error and the build will stop (`set -ex`) 

It may also pay to have a more descriptive name, "gh-actions-conan-ccache-cache" (or whatever takes your fancy) given the machines may be used for other projects?

tequdev replied:
If the volume already exists, it will be ignored.

https://docs.docker.com/reference/cli/docker/volume/create/#:~:text=If%20you%20specify%20a%20volume%20name%20already%20in%20use%20on%20the%20current%20driver%2C%20Docker%20assumes%20you%20want%20to%20reuse%20the%20existing%20volume%20and%20doesn%27t%20return%20an%20error.

There is no objection to using a different name for the volume name. Considering that it will also be used locally, how about something like `xahau-release-builder-cache`?

sublimator replied:
> If the volume already exists, it will be ignored.

Oh :) 

> Considering that it will also be used locally, how about something like xahau-release-builder-cache?

works for me

sublimator commented on build-core.sh:
```diff
@@ -38,12 +37,21 @@ target_link_libraries (ripple_libs INTERFACE wasmedge)
-cmake .. -DCMAKE_BUILD_TYPE=Release -DBoost_NO_BOOST_CMAKE=ON -DLLVM_DIR=/usr/lib64/llvm13/lib/cmake/llvm/ -DLLVM_LIBRARY_DIR=/usr/lib64/llvm13/lib/ -DWasmEdge_LIB=/usr/local/lib64/libwasmedge.a &&
-make -j$3 VERBOSE=1 &&
-strip -s rippled &&
+conan install .. --output-folder . --build missing --settings build_type=$BUILD_TYPE &&
```
We aren't using the soci/snappy from externals which I assume are there for reason, though that brings up questions about cache volume contention. 

See related: #546

tequdev replied:
I’ll add two `conan export` lines.

sublimator commented on release-builder.sh:
```diff
@@ -50,25 +49,159 @@ STATIC_CONTAINER=$(docker ps -a | grep $CONTAINER_NAME |wc -l)
+  DOCKERFILE_CONTENT=$(cat <<'DOCKERFILE_EOF'
+FROM ghcr.io/phusion/holy-build-box:4.0.1-amd64
+
+ARG BUILD_CORES=8
```
It's not really a change here, but "while you're at it", it would be good if this script honoured the BUILD_CORES env, currently if you're not running as a Github Action it's hard coded to 8, which is very wasteful on bigger machines. 

```
BUILD_CORES=$(echo "scale=0 ; `nproc` / 1.337" | bc)

if [[ "$GITHUB_REPOSITORY" == "" ]]; then
  #Default
  BUILD_CORES=8
fi
```

sublimator replied:
https://github.com/tequdev/xahaud/pull/5/files

take it or leave it :) 

I just want ability to set BUILD_CORES plz

tequdev replied:
This setting respects the existing release-builder.sh processing.
How about removing the BUILD_CORE setting processing for cases that are not GitHub Actions?

If you want to build with any number of cores, you can replace the line `BUILD_CORES=$(echo “scale=0 ; `nproc` / 1.337” | bc)`.

sublimator replied:
I'm not really fussed however you do it

I just built on an ubuntu 24.04 machine with many more cores, and didn't realize until a bit of the way through

sublimator replied:
> even BUILD_CORES=${BUILD_CORES:-8}
 
Bamn, build cores as arg destroys docker caching layer btw :/

sublimator commented on build-core.sh:
```diff
@@ -38,12 +37,21 @@ target_link_libraries (ripple_libs INTERFACE wasmedge)
-cmake .. -DCMAKE_BUILD_TYPE=Release -DBoost_NO_BOOST_CMAKE=ON -DLLVM_DIR=/usr/lib64/llvm13/lib/cmake/llvm/ -DLLVM_LIBRARY_DIR=/usr/lib64/llvm13/lib/ -DWasmEdge_LIB=/usr/local/lib64/libwasmedge.a &&
-make -j$3 VERBOSE=1 &&
-strip -s rippled &&
+conan install .. --output-folder . --build missing --settings build_type=$BUILD_TYPE &&
```
reminder: conan has issues with concurrent use of cache

sublimator replied:
I actually had to use "conan remove --locks" at one point when experimenting after canceling a build midway through

sublimator commented on build-core.sh:
```diff
@@ -38,12 +37,21 @@ target_link_libraries (ripple_libs INTERFACE wasmedge)
+
+export LDFLAGS="-static-libstdc++"
+export CMAKE_EXE_LINKER_FLAGS="-static-libstdc++"
+export CMAKE_STATIC_LINKER_FLAGS="-static-libstdc++"
```
According to Gemini, the Conan toolchain file (conan_toolchain.cmake) takes precedence and defines the build environment for CMake. Since those variables aren't explicitly passed into the CMake
  configuration (via the -D flag), they are effectively ignored for the final executable build.

```
 ldd release-build/xahaud       
        linux-vdso.so.1 (0x00007ffc1c1de000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007bebc6c00000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007bebc703d000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007bebc7038000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007bebc7033000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007bebc6f4a000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007bebc6800000)
        /lib64/ld-linux-x86-64.so.2 (0x00007bebc704b000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007bebc6f1a000)
```

sublimator replied:
```
cmake .. \
    -G Ninja \
    -DCMAKE_BUILD_TYPE=$BUILD_TYPE \
    -DCMAKE_TOOLCHAIN_FILE:FILEPATH=build/generators/conan_toolchain.cmake \
    -DCMAKE_EXE_LINKER_FLAGS="-static-libstdc++" \
    -DCMAKE_STATIC_LINKER_FLAGS="-static-libstdc++" \
    -DLLVM_DIR=$LLVM_DIR \
    -DWasmEdge_LIB=$WasmEdge_LIB \
    -Dxrpld=TRUE \
    -Dtests=TRUE &&
```

sublimator replied:
```
[62/673] Building CXX object CMakeFiles/xrpl_core.dir/src/ripple/protocol/impl/TER.cpp.o
[63/673] Linking C static library src/ed25519-donna/libed25519.a
FAILED: src/ed25519-donna/libed25519.a 
: && /usr/bin/cmake -E rm -f src/ed25519-donna/libed25519.a && /usr/bin/ar qc src/ed25519-donna/libed25519.a -static-libstdc++ src/ed25519-donna/CMakeFiles/ed25519.dir/ed25519.c.o && /usr/bin/ranlib src/ed25519-donna/libed25519.a && :
/usr/bin/ar: two different operation options specified
[64/673] Building CXX object CMakeFiles/xrpl_core.dir/src/ripple/crypto/impl/RFC1751.cpp.o
[65/673] Building CXX object CMakeFiles/xrpl_core.dir/src/ripple/protocol/impl/PublicKey.cpp.o
```

Oh, wait

```
-DCMAKE_EXE_LINKER_FLAGS="-static-libstdc++" \
-DCMAKE_STATIC_LINKER_FLAGS="-static-libstdc++" \
```

***

### CMAKE_EXE_LINKER_FLAGS

* **Tool:** The **linker** (like `ld`, which is usually called by the compiler driver `g++` or `clang++`).
* **Purpose:** To link object files and libraries together to create a final executable or a shared library (`.so`).
* **Correct Usage:** This is the correct place to put linker-specific flags like `-static-libstdc++`.

***

### CMAKE_STATIC_LINKER_FLAGS

* **Tool:** The **archiver** (`ar`).
* **Purpose:** To bundle object files (`.o`) into a static library archive (`.a`). It doesn't perform linking; it just collects files into a single package.
* **Correct Usage:** This variable should **not** contain linker flags. Passing `-static-libstdc++` here is what caused your error because `ar` doesn't understand that flag.

In short, you've diagnosed the problem correctly: a flag meant for the linker was incorrectly passed to the archiver. Moving `-static-libstdc++` from `CMAKE_STATIC_LINKER_FLAGS` to `CMAKE_EXE_LINKER_FLAGS` should resolve the build failure.

sublimator replied:
set_target_properties(rippled PROPERTIES LINK_FLAGS "-static-libstdc++")

Though, it seems we might actually need this in RippledCore.cmake, because `CMAKE_EXE_LINKER_FLAGS` is not being honoured

sublimator replied:
Nor is that, actually

sublimator replied:
Suspecting conan binaries linked against dynamic libstdc++.so.6

sublimator replied:
Seems to be snappy 

```
> ╭─niq@niq-xahau ~/xahau/xahaud-worktrees/xahaud/release-build/build/generators ‹conan-release-builder●› 
  ╰─$ grep -r -- 'stdc++'   
  Snappy-release-x86_64-data.cmake:set(snappy_SYSTEM_LIBS_RELEASE m stdc++)
  Snappy-release-x86_64-data.cmake:set(snappy_Snappy_snappy_SYSTEM_LIBS_RELEASE m stdc++)
  ╭─niq@niq-xahau ~/xahau/xahaud-worktrees/xahaud/release-build/build/generators ‹conan-release-builder●›
```

sublimator replied:
https://gist.github.com/sublimator/23cdb66ca497139fed256e559d9e75e4

sublimator replied:
Let me correct myself and stick only to the hard evidence we have gathered.

  Here are the facts, without any assumptions:

  1. Fact: The sed command that removes stdc++ from Snappy-release-x86_64-data.cmake works. It produces a correctly linked binary. This is proven by your final ldd output.
  2. Fact: The grep command proves that Snappy-release-x86_64-data.cmake is the only file in the entire build directory that contains the incorrect stdc++ system library directive.
  3. Fact: The snappy/conanfile.py is the only recipe we have inspected that explicitly adds a library ("m") to the system_libs property.
  4. Fact: All attempts to fix this by changing global settings (like the Conan profile or command-line flags) have failed.

  Based on these facts, the only logical conclusion is that there is a unique, specific interaction between the snappy recipe and the Conan generator that causes the incorrect file to
   be created.

  You are right to say that my explanation of why the generator does this is an assumption. But we don't need to know the exact "why" to be certain of the "what" and the "where."

  The evidence proves the problem is isolated to the generation of that one file, and therefore, your sed patch is the correct, targeted, and evidence-based solution.

sublimator replied:
I removed completely the `system_libs.append("m")` from the snappy recipe, because `m` was always going to get dragged into the mess, in a system as complex as `{xahau,ripple}d` and it confirms the theory of the `m` triggering some accidental inclusion of `lstdc++`. 

```
> ╭─niq@niq-xahau ~/xahau/xahaud-worktrees/xahaud ‹conan-release-builder●› 
  ╰─$ ldd release-build/xahaud                                   
          linux-vdso.so.1 (0x00007ffd7119f000)
          libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x0000719157dda000)
          librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x0000719157dd5000)
          libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x0000719157dd0000)
          libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x0000719157ce7000)
          libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x0000719157a00000)
          /lib64/ld-linux-x86-64.so.2 (0x0000719157de8000)
```

We could just comment it out in our external, add comments as to why, and if we were feeling pedantic enough, explicitly add `m` to our executable targets.

sublimator replied:
For record:
https://github.com/tequdev/xahaud/pull/6/files

tequdev committed:
add conan export snapply,soci

tequdev committed:
fix volume name, BUILD_CORES

tequdev committed:
fix soci boost option

sublimator commented on build-core.sh:
```diff
@@ -38,12 +37,23 @@ target_link_libraries (ripple_libs INTERFACE wasmedge)
 git checkout src/ripple/protocol/impl/BuildInfo.cpp &&
-sed -i s/\"0.0.0\"/\"$(date +%Y).$(date +%-m).$(date +%-d)-$(git rev-parse --abbrev-ref HEAD)+$4\"/g src/ripple/protocol/impl/BuildInfo.cpp &&
+sed -i s/\"0.0.0\"/\"$(date +%Y).$(date +%-m).$(date +%-d)-$(git rev-parse --abbrev-ref HEAD)$(if [ -n "$4" ]; then echo "+$4"; fi)\"/g src/ripple/protocol/impl/BuildInfo.cpp &&
+conan export external/snappy snappy/1.1.9@ &&
```
See #546 
Our conanfile is using 1.1.10 
And the fact that we mispecified this again, is a really good argument for namespacing! 

```
conan export external/snappy snappy/1.1.10@xahaud/stable
conan export external/soci soci/4.0.3@xahaud/stable
```

I don't know about conan 2, but with conan 1, when we are exporting with only "soci/4.0.3@" and our recipe is different to rippled, which also exports using that name, you end up having to export again every time you switch between xahaud and rippled.

Also, as you can see, it's very easy for the builds to silently use the conan version.

tequdev replied:
I'll fix this to 1.1.10.

and I'll fix the namespace once the PR is merged.

sublimator committed:
fix(build): ensure static libstdc++ linking by patching conan deps

This commit resolves an issue where the xahaud binary was dynamically
linking against libstdc++ despite build flags requesting static linkage.

The root cause was a bug in the Conan 1.x CMakeDeps generator,
triggered by the snappy recipe modifying `system_libs`. This caused
Conan to inject `-lstdc++` into the build, overriding other settings.

The fix involves three parts:
1. snappy/conanfile.py: The trigger for the bug is removed by
   commenting out the modification of `system_libs`.
2. RippledCore.cmake: The math library (`libm`) is now explicitly
   linked to the final executable target as a best practice.
3. build-core.sh: The conan export for snappy is updated to the
   correct version.

tequdev committed:
Merge commit 'e192deec3aa6f156a8da7542c55becc1ae80706e' into conan-release-builder

tequdev committed:
add conan settings.compiler.libcxx

tequdev committed:
add binary libcheck

tequdev committed:
fix CMAKE_EXE_LINKER_FLAGS use

tequdev committed:
fix conan install fail

sublimator commented on build-core.sh:
```diff
@@ -55,6 +55,7 @@ ninja -j $3 &&
 ccache -s &&
 strip -s rippled &&	
 mv rippled xahaud &&
+libcheck xahaud &&
```
For anyone else curious:
https://github.com/phusion/holy-build-box/blob/master/image/libcheck

sublimator commented on build-core.sh:
```diff
@@ -38,13 +37,30 @@ target_link_libraries (ripple_libs INTERFACE wasmedge)
 message(\"WasmEdge DONE\")
 " > Builds/CMake/deps/WasmEdge.cmake &&
+
+export LDFLAGS="-static-libstdc++"
```
It seems to be working, but we should probably be "extending" LDFLAGS:
https://github.com/phusion/holy-build-box/blob/master/image/activate_func.sh#L14-L15

```
export LDFLAGS="$LDFLAGS -static-libstdc++"
```

sublimator replied:
activate_holy_build_box_deps_installation_environment

Though that may not be relevant at all

sublimator replied:
https://github.com/phusion/holy-build-box/blob/master/image/activate_func.sh#L24-L45

Ok, no, `activate_holy_build_box` is setting LDFLAGS too

sublimator replied:
also, comment that is relevant to the conan deps (my assumption) probably good

RichardAH commented:
Awesome job guys
